lib/charms/nginx_ingress_integrator/v0/ingress.py

```diff
@@ -189,40 +249,76 @@ def _on_relation_changed(self, event):
         )
 
         if missing_fields:
-            logger.error(
-                "Missing required data fields for ingress relation: {}".format(
-                    ", ".join(missing_fields)
-                )
+            LOGGER.error(
+                "Missing required data fields for ingress relation: %s",
+                ", ".join(missing_fields),
             )
             self.model.unit.status = BlockedStatus(
-                "Missing fields for ingress: {}".format(", ".join(missing_fields))
+                f"Missing fields for ingress: {', '.join(missing_fields)}"
             )
 
+        # Conform to charm-relation-interfaces.
+        if "name" in ingress_data and "port" in ingress_data:
+            name = ingress_data["name"]
+            port = ingress_data["port"]
+        else:
+            name = ingress_data["service-name"]
+            port = ingress_data["service-port"]
+        event.relation.data[self.model.app]["url"] = f"http://{name}:{port}/"
+
         # Create an event that our charm can use to decide it's okay to
         # configure the ingress.
         self.charm.on.ingress_available.emit()
 
+    def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Handle a relation-broken event in the ingress relation."""
+        if not self.model.unit.is_leader():
+            return
 
-class IngressProvides(IngressBaseProvides):
-    """This class defines the functionality for the 'provides' side of the 'ingress' relation.
+        # Create an event that our charm can use to remove the ingress resource.
+        self.charm.on.ingress_broken.emit(event.relation)
 
-    Hook events observed:
-        - relation-changed
-    """
+class IngressProxyProvides(Object):
 
     def __init__(self, charm):
-        super().__init__(charm, "ingress")
+        super().__init__(charm, "ingress-proxy")
         # Observe the relation-changed hook event and bind
         # self.on_relation_changed() to handle the event.
-        self.framework.observe(charm.on["ingress"].relation_changed, self._on_relation_changed)
+        self.framework.observe(charm.on["ingress-proxy"].relation_changed, self._on_relation_changed)
         self.charm = charm
 
+    def _on_relation_changed(self, event):
```

Comment:
There seems to be a lot of duplication here between this and the _on_relation_changed method for IngressProvides. How about if we add a `relation_name` (or similar) parameter to that method which allows us to share between the two. Or maybe we can inspect the relation name that's being passed to avoid needing to pass a different parameter at all?